// codelyzer/util/syntaxKind.js
"use strict";
var ts = require("typescript");
var SyntaxKind;
(function (SyntaxKind) {
    var cachedKinds = null;
    var SyntaxKind_1_5 = {
        Unknown: 0,
        EndOfFileToken: 1,
        SingleLineCommentTrivia: 2,
        MultiLineCommentTrivia: 3,
        NewLineTrivia: 4,
        WhitespaceTrivia: 5,
        ConflictMarkerTrivia: 6,
        NumericLiteral: 7,
        StringLiteral: 8,
        RegularExpressionLiteral: 9,
        NoSubstitutionTemplateLiteral: 10,
        TemplateHead: 11,
        TemplateMiddle: 12,
        TemplateTail: 13,
        OpenBraceToken: 14,
        CloseBraceToken: 15,
        OpenParenToken: 16,
        CloseParenToken: 17,
        OpenBracketToken: 18,
        CloseBracketToken: 19,
        DotToken: 20,
        DotDotDotToken: 21,
        SemicolonToken: 22,
        CommaToken: 23,
        LessThanToken: 24,
        GreaterThanToken: 25,
        LessThanEqualsToken: 26,
        GreaterThanEqualsToken: 27,
        EqualsEqualsToken: 28,
        ExclamationEqualsToken: 29,
        EqualsEqualsEqualsToken: 30,
        ExclamationEqualsEqualsToken: 31,
        EqualsGreaterThanToken: 32,
        PlusToken: 33,
        MinusToken: 34,
        AsteriskToken: 35,
        SlashToken: 36,
        PercentToken: 37,
        PlusPlusToken: 38,
        MinusMinusToken: 39,
        LessThanLessThanToken: 40,
        GreaterThanGreaterThanToken: 41,
        GreaterThanGreaterThanGreaterThanToken: 42,
        AmpersandToken: 43,
        BarToken: 44,
        CaretToken: 45,
        ExclamationToken: 46,
        TildeToken: 47,
        AmpersandAmpersandToken: 48,
        BarBarToken: 49,
        QuestionToken: 50,
        ColonToken: 51,
        AtToken: 52,
        EqualsToken: 53,
        PlusEqualsToken: 54,
        MinusEqualsToken: 55,
        AsteriskEqualsToken: 56,
        SlashEqualsToken: 57,
        PercentEqualsToken: 58,
        LessThanLessThanEqualsToken: 59,
        GreaterThanGreaterThanEqualsToken: 60,
        GreaterThanGreaterThanGreaterThanEqualsToken: 61,
        AmpersandEqualsToken: 62,
        BarEqualsToken: 63,
        CaretEqualsToken: 64,
        Identifier: 65,
        BreakKeyword: 66,
        CaseKeyword: 67,
        CatchKeyword: 68,
        ClassKeyword: 69,
        ConstKeyword: 70,
        ContinueKeyword: 71,
        DebuggerKeyword: 72,
        DefaultKeyword: 73,
        DeleteKeyword: 74,
        DoKeyword: 75,
        ElseKeyword: 76,
        EnumKeyword: 77,
        ExportKeyword: 78,
        ExtendsKeyword: 79,
        FalseKeyword: 80,
        FinallyKeyword: 81,
        ForKeyword: 82,
        FunctionKeyword: 83,
        IfKeyword: 84,
        ImportKeyword: 85,
        InKeyword: 86,
        InstanceOfKeyword: 87,
        NewKeyword: 88,
        NullKeyword: 89,
        ReturnKeyword: 90,
        SuperKeyword: 91,
        SwitchKeyword: 92,
        ThisKeyword: 93,
        ThrowKeyword: 94,
        TrueKeyword: 95,
        TryKeyword: 96,
        TypeOfKeyword: 97,
        VarKeyword: 98,
        VoidKeyword: 99,
        WhileKeyword: 100,
        WithKeyword: 101,
        ImplementsKeyword: 102,
        InterfaceKeyword: 103,
        LetKeyword: 104,
        PackageKeyword: 105,
        PrivateKeyword: 106,
        ProtectedKeyword: 107,
        PublicKeyword: 108,
        StaticKeyword: 109,
        YieldKeyword: 110,
        AsKeyword: 111,
        AnyKeyword: 112,
        BooleanKeyword: 113,
        ConstructorKeyword: 114,
        DeclareKeyword: 115,
        GetKeyword: 116,
        ModuleKeyword: 117,
        NamespaceKeyword: 118,
        RequireKeyword: 119,
        NumberKeyword: 120,
        SetKeyword: 121,
        StringKeyword: 122,
        SymbolKeyword: 123,
        TypeKeyword: 124,
        FromKeyword: 125,
        OfKeyword: 126,
        QualifiedName: 127,
        ComputedPropertyName: 128,
        TypeParameter: 129,
        Parameter: 130,
        Decorator: 131,
        PropertySignature: 132,
        PropertyDeclaration: 133,
        MethodSignature: 134,
        MethodDeclaration: 135,
        Constructor: 136,
        GetAccessor: 137,
        SetAccessor: 138,
        CallSignature: 139,
        ConstructSignature: 140,
        IndexSignature: 141,
        TypeReference: 142,
        FunctionType: 143,
        ConstructorType: 144,
        TypeQuery: 145,
        TypeLiteral: 146,
        ArrayType: 147,
        TupleType: 148,
        UnionType: 149,
        ParenthesizedType: 150,
        ObjectBindingPattern: 151,
        ArrayBindingPattern: 152,
        BindingElement: 153,
        ArrayLiteralExpression: 154,
        ObjectLiteralExpression: 155,
        PropertyAccessExpression: 156,
        ElementAccessExpression: 157,
        CallExpression: 158,
        NewExpression: 159,
        TaggedTemplateExpression: 160,
        TypeAssertionExpression: 161,
        ParenthesizedExpression: 162,
        FunctionExpression: 163,
        ArrowFunction: 164,
        DeleteExpression: 165,
        TypeOfExpression: 166,
        VoidExpression: 167,
        PrefixUnaryExpression: 168,
        PostfixUnaryExpression: 169,
        BinaryExpression: 170,
        ConditionalExpression: 171,
        TemplateExpression: 172,
        YieldExpression: 173,
        SpreadElementExpression: 174,
        ClassExpression: 175,
        OmittedExpression: 176,
        ExpressionWithTypeArguments: 177,
        TemplateSpan: 178,
        SemicolonClassElement: 179,
        Block: 180,
        VariableStatement: 181,
        EmptyStatement: 182,
        ExpressionStatement: 183,
        IfStatement: 184,
        DoStatement: 185,
        WhileStatement: 186,
        ForStatement: 187,
        ForInStatement: 188,
        ForOfStatement: 189,
        ContinueStatement: 190,
        BreakStatement: 191,
        ReturnStatement: 192,
        WithStatement: 193,
        SwitchStatement: 194,
        LabeledStatement: 195,
        ThrowStatement: 196,
        TryStatement: 197,
        DebuggerStatement: 198,
        VariableDeclaration: 199,
        VariableDeclarationList: 200,
        FunctionDeclaration: 201,
        ClassDeclaration: 202,
        InterfaceDeclaration: 203,
        TypeAliasDeclaration: 204,
        EnumDeclaration: 205,
        ModuleDeclaration: 206,
        ModuleBlock: 207,
        CaseBlock: 208,
        ImportEqualsDeclaration: 209,
        ImportDeclaration: 210,
        ImportClause: 211,
        NamespaceImport: 212,
        NamedImports: 213,
        ImportSpecifier: 214,
        ExportAssignment: 215,
        ExportDeclaration: 216,
        NamedExports: 217,
        ExportSpecifier: 218,
        MissingDeclaration: 219,
        ExternalModuleReference: 220,
        CaseClause: 221,
        DefaultClause: 222,
        HeritageClause: 223,
        CatchClause: 224,
        PropertyAssignment: 225,
        ShorthandPropertyAssignment: 226,
        EnumMember: 227,
        SourceFile: 228,
        SyntaxList: 229,
        Count: 230,
        FirstAssignment: 53,
        LastAssignment: 64,
        FirstReservedWord: 66,
        LastReservedWord: 101,
        FirstKeyword: 66,
        LastKeyword: 126,
        FirstFutureReservedWord: 102,
        LastFutureReservedWord: 110,
        FirstTypeNode: 142,
        LastTypeNode: 150,
        FirstPunctuation: 14,
        LastPunctuation: 64,
        FirstToken: 0,
        LastToken: 126,
        FirstTriviaToken: 2,
        LastTriviaToken: 6,
        FirstLiteralToken: 7,
        LastLiteralToken: 10,
        FirstTemplateToken: 10,
        LastTemplateToken: 13,
        FirstBinaryOperator: 24,
        LastBinaryOperator: 64,
        FirstNode: 127
    };
    var SyntaxKind_1_6 = {
        Unknown: 0,
        EndOfFileToken: 1,
        SingleLineCommentTrivia: 2,
        MultiLineCommentTrivia: 3,
        NewLineTrivia: 4,
        WhitespaceTrivia: 5,
        ShebangTrivia: 6,
        ConflictMarkerTrivia: 7,
        NumericLiteral: 8,
        StringLiteral: 9,
        RegularExpressionLiteral: 10,
        NoSubstitutionTemplateLiteral: 11,
        TemplateHead: 12,
        TemplateMiddle: 13,
        TemplateTail: 14,
        OpenBraceToken: 15,
        CloseBraceToken: 16,
        OpenParenToken: 17,
        CloseParenToken: 18,
        OpenBracketToken: 19,
        CloseBracketToken: 20,
        DotToken: 21,
        DotDotDotToken: 22,
        SemicolonToken: 23,
        CommaToken: 24,
        LessThanToken: 25,
        LessThanSlashToken: 26,
        GreaterThanToken: 27,
        LessThanEqualsToken: 28,
        GreaterThanEqualsToken: 29,
        EqualsEqualsToken: 30,
        ExclamationEqualsToken: 31,
        EqualsEqualsEqualsToken: 32,
        ExclamationEqualsEqualsToken: 33,
        EqualsGreaterThanToken: 34,
        PlusToken: 35,
        MinusToken: 36,
        AsteriskToken: 37,
        SlashToken: 38,
        PercentToken: 39,
        PlusPlusToken: 40,
        MinusMinusToken: 41,
        LessThanLessThanToken: 42,
        GreaterThanGreaterThanToken: 43,
        GreaterThanGreaterThanGreaterThanToken: 44,
        AmpersandToken: 45,
        BarToken: 46,
        CaretToken: 47,
        ExclamationToken: 48,
        TildeToken: 49,
        AmpersandAmpersandToken: 50,
        BarBarToken: 51,
        QuestionToken: 52,
        ColonToken: 53,
        AtToken: 54,
        EqualsToken: 55,
        PlusEqualsToken: 56,
        MinusEqualsToken: 57,
        AsteriskEqualsToken: 58,
        SlashEqualsToken: 59,
        PercentEqualsToken: 60,
        LessThanLessThanEqualsToken: 61,
        GreaterThanGreaterThanEqualsToken: 62,
        GreaterThanGreaterThanGreaterThanEqualsToken: 63,
        AmpersandEqualsToken: 64,
        BarEqualsToken: 65,
        CaretEqualsToken: 66,
        Identifier: 67,
        BreakKeyword: 68,
        CaseKeyword: 69,
        CatchKeyword: 70,
        ClassKeyword: 71,
        ConstKeyword: 72,
        ContinueKeyword: 73,
        DebuggerKeyword: 74,
        DefaultKeyword: 75,
        DeleteKeyword: 76,
        DoKeyword: 77,
        ElseKeyword: 78,
        EnumKeyword: 79,
        ExportKeyword: 80,
        ExtendsKeyword: 81,
        FalseKeyword: 82,
        FinallyKeyword: 83,
        ForKeyword: 84,
        FunctionKeyword: 85,
        IfKeyword: 86,
        ImportKeyword: 87,
        InKeyword: 88,
        InstanceOfKeyword: 89,
        NewKeyword: 90,
        NullKeyword: 91,
        ReturnKeyword: 92,
        SuperKeyword: 93,
        SwitchKeyword: 94,
        ThisKeyword: 95,
        ThrowKeyword: 96,
        TrueKeyword: 97,
        TryKeyword: 98,
        TypeOfKeyword: 99,
        VarKeyword: 100,
        VoidKeyword: 101,
        WhileKeyword: 102,
        WithKeyword: 103,
        ImplementsKeyword: 104,
        InterfaceKeyword: 105,
        LetKeyword: 106,
        PackageKeyword: 107,
        PrivateKeyword: 108,
        ProtectedKeyword: 109,
        PublicKeyword: 110,
        StaticKeyword: 111,
        YieldKeyword: 112,
        AbstractKeyword: 113,
        AsKeyword: 114,
        AnyKeyword: 115,
        AsyncKeyword: 116,
        AwaitKeyword: 117,
        BooleanKeyword: 118,
        ConstructorKeyword: 119,
        DeclareKeyword: 120,
        GetKeyword: 121,
        IsKeyword: 122,
        ModuleKeyword: 123,
        NamespaceKeyword: 124,
        RequireKeyword: 125,
        NumberKeyword: 126,
        SetKeyword: 127,
        StringKeyword: 128,
        SymbolKeyword: 129,
        TypeKeyword: 130,
        FromKeyword: 131,
        OfKeyword: 132,
        QualifiedName: 133,
        ComputedPropertyName: 134,
        TypeParameter: 135,
        Parameter: 136,
        Decorator: 137,
        PropertySignature: 138,
        PropertyDeclaration: 139,
        MethodSignature: 140,
        MethodDeclaration: 141,
        Constructor: 142,
        GetAccessor: 143,
        SetAccessor: 144,
        CallSignature: 145,
        ConstructSignature: 146,
        IndexSignature: 147,
        TypePredicate: 148,
        TypeReference: 149,
        FunctionType: 150,
        ConstructorType: 151,
        TypeQuery: 152,
        TypeLiteral: 153,
        ArrayType: 154,
        TupleType: 155,
        UnionType: 156,
        IntersectionType: 157,
        ParenthesizedType: 158,
        ObjectBindingPattern: 159,
        ArrayBindingPattern: 160,
        BindingElement: 161,
        ArrayLiteralExpression: 162,
        ObjectLiteralExpression: 163,
        PropertyAccessExpression: 164,
        ElementAccessExpression: 165,
        CallExpression: 166,
        NewExpression: 167,
        TaggedTemplateExpression: 168,
        TypeAssertionExpression: 169,
        ParenthesizedExpression: 170,
        FunctionExpression: 171,
        ArrowFunction: 172,
        DeleteExpression: 173,
        TypeOfExpression: 174,
        VoidExpression: 175,
        AwaitExpression: 176,
        PrefixUnaryExpression: 177,
        PostfixUnaryExpression: 178,
        BinaryExpression: 179,
        ConditionalExpression: 180,
        TemplateExpression: 181,
        YieldExpression: 182,
        SpreadElementExpression: 183,
        ClassExpression: 184,
        OmittedExpression: 185,
        ExpressionWithTypeArguments: 186,
        AsExpression: 187,
        TemplateSpan: 188,
        SemicolonClassElement: 189,
        Block: 190,
        VariableStatement: 191,
        EmptyStatement: 192,
        ExpressionStatement: 193,
        IfStatement: 194,
        DoStatement: 195,
        WhileStatement: 196,
        ForStatement: 197,
        ForInStatement: 198,
        ForOfStatement: 199,
        ContinueStatement: 200,
        BreakStatement: 201,
        ReturnStatement: 202,
        WithStatement: 203,
        SwitchStatement: 204,
        LabeledStatement: 205,
        ThrowStatement: 206,
        TryStatement: 207,
        DebuggerStatement: 208,
        VariableDeclaration: 209,
        VariableDeclarationList: 210,
        FunctionDeclaration: 211,
        ClassDeclaration: 212,
        InterfaceDeclaration: 213,
        TypeAliasDeclaration: 214,
        EnumDeclaration: 215,
        ModuleDeclaration: 216,
        ModuleBlock: 217,
        CaseBlock: 218,
        ImportEqualsDeclaration: 219,
        ImportDeclaration: 220,
        ImportClause: 221,
        NamespaceImport: 222,
        NamedImports: 223,
        ImportSpecifier: 224,
        ExportAssignment: 225,
        ExportDeclaration: 226,
        NamedExports: 227,
        ExportSpecifier: 228,
        MissingDeclaration: 229,
        ExternalModuleReference: 230,
        JsxElement: 231,
        JsxSelfClosingElement: 232,
        JsxOpeningElement: 233,
        JsxText: 234,
        JsxClosingElement: 235,
        JsxAttribute: 236,
        JsxSpreadAttribute: 237,
        JsxExpression: 238,
        CaseClause: 239,
        DefaultClause: 240,
        HeritageClause: 241,
        CatchClause: 242,
        PropertyAssignment: 243,
        ShorthandPropertyAssignment: 244,
        EnumMember: 245,
        SourceFile: 246,
        JSDocTypeExpression: 247,
        JSDocAllType: 248,
        JSDocUnknownType: 249,
        JSDocArrayType: 250,
        JSDocUnionType: 251,
        JSDocTupleType: 252,
        JSDocNullableType: 253,
        JSDocNonNullableType: 254,
        JSDocRecordType: 255,
        JSDocRecordMember: 256,
        JSDocTypeReference: 257,
        JSDocOptionalType: 258,
        JSDocFunctionType: 259,
        JSDocVariadicType: 260,
        JSDocConstructorType: 261,
        JSDocThisType: 262,
        JSDocComment: 263,
        JSDocTag: 264,
        JSDocParameterTag: 265,
        JSDocReturnTag: 266,
        JSDocTypeTag: 267,
        JSDocTemplateTag: 268,
        SyntaxList: 269,
        Count: 270,
        FirstAssignment: 55,
        LastAssignment: 66,
        FirstReservedWord: 68,
        LastReservedWord: 103,
        FirstKeyword: 68,
        LastKeyword: 132,
        FirstFutureReservedWord: 104,
        LastFutureReservedWord: 112,
        FirstTypeNode: 149,
        LastTypeNode: 158,
        FirstPunctuation: 15,
        LastPunctuation: 66,
        FirstToken: 0,
        LastToken: 132,
        FirstTriviaToken: 2,
        LastTriviaToken: 7,
        FirstLiteralToken: 8,
        LastLiteralToken: 11,
        FirstTemplateToken: 11,
        LastTemplateToken: 14,
        FirstBinaryOperator: 25,
        LastBinaryOperator: 66,
        FirstNode: 133
    };
    var SyntaxKind_1_7 = {
        Unknown: 0,
        EndOfFileToken: 1,
        SingleLineCommentTrivia: 2,
        MultiLineCommentTrivia: 3,
        NewLineTrivia: 4,
        WhitespaceTrivia: 5,
        ShebangTrivia: 6,
        ConflictMarkerTrivia: 7,
        NumericLiteral: 8,
        StringLiteral: 9,
        RegularExpressionLiteral: 10,
        NoSubstitutionTemplateLiteral: 11,
        TemplateHead: 12,
        TemplateMiddle: 13,
        TemplateTail: 14,
        OpenBraceToken: 15,
        CloseBraceToken: 16,
        OpenParenToken: 17,
        CloseParenToken: 18,
        OpenBracketToken: 19,
        CloseBracketToken: 20,
        DotToken: 21,
        DotDotDotToken: 22,
        SemicolonToken: 23,
        CommaToken: 24,
        LessThanToken: 25,
        LessThanSlashToken: 26,
        GreaterThanToken: 27,
        LessThanEqualsToken: 28,
        GreaterThanEqualsToken: 29,
        EqualsEqualsToken: 30,
        ExclamationEqualsToken: 31,
        EqualsEqualsEqualsToken: 32,
        ExclamationEqualsEqualsToken: 33,
        EqualsGreaterThanToken: 34,
        PlusToken: 35,
        MinusToken: 36,
        AsteriskToken: 37,
        AsteriskAsteriskToken: 38,
        SlashToken: 39,
        PercentToken: 40,
        PlusPlusToken: 41,
        MinusMinusToken: 42,
        LessThanLessThanToken: 43,
        GreaterThanGreaterThanToken: 44,
        GreaterThanGreaterThanGreaterThanToken: 45,
        AmpersandToken: 46,
        BarToken: 47,
        CaretToken: 48,
        ExclamationToken: 49,
        TildeToken: 50,
        AmpersandAmpersandToken: 51,
        BarBarToken: 52,
        QuestionToken: 53,
        ColonToken: 54,
        AtToken: 55,
        EqualsToken: 56,
        PlusEqualsToken: 57,
        MinusEqualsToken: 58,
        AsteriskEqualsToken: 59,
        AsteriskAsteriskEqualsToken: 60,
        SlashEqualsToken: 61,
        PercentEqualsToken: 62,
        LessThanLessThanEqualsToken: 63,
        GreaterThanGreaterThanEqualsToken: 64,
        GreaterThanGreaterThanGreaterThanEqualsToken: 65,
        AmpersandEqualsToken: 66,
        BarEqualsToken: 67,
        CaretEqualsToken: 68,
        Identifier: 69,
        BreakKeyword: 70,
        CaseKeyword: 71,
        CatchKeyword: 72,
        ClassKeyword: 73,
        ConstKeyword: 74,
        ContinueKeyword: 75,
        DebuggerKeyword: 76,
        DefaultKeyword: 77,
        DeleteKeyword: 78,
        DoKeyword: 79,
        ElseKeyword: 80,
        EnumKeyword: 81,
        ExportKeyword: 82,
        ExtendsKeyword: 83,
        FalseKeyword: 84,
        FinallyKeyword: 85,
        ForKeyword: 86,
        FunctionKeyword: 87,
        IfKeyword: 88,
        ImportKeyword: 89,
        InKeyword: 90,
        InstanceOfKeyword: 91,
        NewKeyword: 92,
        NullKeyword: 93,
        ReturnKeyword: 94,
        SuperKeyword: 95,
        SwitchKeyword: 96,
        ThisKeyword: 97,
        ThrowKeyword: 98,
        TrueKeyword: 99,
        TryKeyword: 100,
        TypeOfKeyword: 101,
        VarKeyword: 102,
        VoidKeyword: 103,
        WhileKeyword: 104,
        WithKeyword: 105,
        ImplementsKeyword: 106,
        InterfaceKeyword: 107,
        LetKeyword: 108,
        PackageKeyword: 109,
        PrivateKeyword: 110,
        ProtectedKeyword: 111,
        PublicKeyword: 112,
        StaticKeyword: 113,
        YieldKeyword: 114,
        AbstractKeyword: 115,
        AsKeyword: 116,
        AnyKeyword: 117,
        AsyncKeyword: 118,
        AwaitKeyword: 119,
        BooleanKeyword: 120,
        ConstructorKeyword: 121,
        DeclareKeyword: 122,
        GetKeyword: 123,
        IsKeyword: 124,
        ModuleKeyword: 125,
        NamespaceKeyword: 126,
        RequireKeyword: 127,
        NumberKeyword: 128,
        SetKeyword: 129,
        StringKeyword: 130,
        SymbolKeyword: 131,
        TypeKeyword: 132,
        FromKeyword: 133,
        OfKeyword: 134,
        QualifiedName: 135,
        ComputedPropertyName: 136,
        TypeParameter: 137,
        Parameter: 138,
        Decorator: 139,
        PropertySignature: 140,
        PropertyDeclaration: 141,
        MethodSignature: 142,
        MethodDeclaration: 143,
        Constructor: 144,
        GetAccessor: 145,
        SetAccessor: 146,
        CallSignature: 147,
        ConstructSignature: 148,
        IndexSignature: 149,
        TypePredicate: 150,
        TypeReference: 151,
        FunctionType: 152,
        ConstructorType: 153,
        TypeQuery: 154,
        TypeLiteral: 155,
        ArrayType: 156,
        TupleType: 157,
        UnionType: 158,
        IntersectionType: 159,
        ParenthesizedType: 160,
        ObjectBindingPattern: 161,
        ArrayBindingPattern: 162,
        BindingElement: 163,
        ArrayLiteralExpression: 164,
        ObjectLiteralExpression: 165,
        PropertyAccessExpression: 166,
        ElementAccessExpression: 167,
        CallExpression: 168,
        NewExpression: 169,
        TaggedTemplateExpression: 170,
        TypeAssertionExpression: 171,
        ParenthesizedExpression: 172,
        FunctionExpression: 173,
        ArrowFunction: 174,
        DeleteExpression: 175,
        TypeOfExpression: 176,
        VoidExpression: 177,
        AwaitExpression: 178,
        PrefixUnaryExpression: 179,
        PostfixUnaryExpression: 180,
        BinaryExpression: 181,
        ConditionalExpression: 182,
        TemplateExpression: 183,
        YieldExpression: 184,
        SpreadElementExpression: 185,
        ClassExpression: 186,
        OmittedExpression: 187,
        ExpressionWithTypeArguments: 188,
        AsExpression: 189,
        TemplateSpan: 190,
        SemicolonClassElement: 191,
        Block: 192,
        VariableStatement: 193,
        EmptyStatement: 194,
        ExpressionStatement: 195,
        IfStatement: 196,
        DoStatement: 197,
        WhileStatement: 198,
        ForStatement: 199,
        ForInStatement: 200,
        ForOfStatement: 201,
        ContinueStatement: 202,
        BreakStatement: 203,
        ReturnStatement: 204,
        WithStatement: 205,
        SwitchStatement: 206,
        LabeledStatement: 207,
        ThrowStatement: 208,
        TryStatement: 209,
        DebuggerStatement: 210,
        VariableDeclaration: 211,
        VariableDeclarationList: 212,
        FunctionDeclaration: 213,
        ClassDeclaration: 214,
        InterfaceDeclaration: 215,
        TypeAliasDeclaration: 216,
        EnumDeclaration: 217,
        ModuleDeclaration: 218,
        ModuleBlock: 219,
        CaseBlock: 220,
        ImportEqualsDeclaration: 221,
        ImportDeclaration: 222,
        ImportClause: 223,
        NamespaceImport: 224,
        NamedImports: 225,
        ImportSpecifier: 226,
        ExportAssignment: 227,
        ExportDeclaration: 228,
        NamedExports: 229,
        ExportSpecifier: 230,
        MissingDeclaration: 231,
        ExternalModuleReference: 232,
        JsxElement: 233,
        JsxSelfClosingElement: 234,
        JsxOpeningElement: 235,
        JsxText: 236,
        JsxClosingElement: 237,
        JsxAttribute: 238,
        JsxSpreadAttribute: 239,
        JsxExpression: 240,
        CaseClause: 241,
        DefaultClause: 242,
        HeritageClause: 243,
        CatchClause: 244,
        PropertyAssignment: 245,
        ShorthandPropertyAssignment: 246,
        EnumMember: 247,
        SourceFile: 248,
        JSDocTypeExpression: 249,
        JSDocAllType: 250,
        JSDocUnknownType: 251,
        JSDocArrayType: 252,
        JSDocUnionType: 253,
        JSDocTupleType: 254,
        JSDocNullableType: 255,
        JSDocNonNullableType: 256,
        JSDocRecordType: 257,
        JSDocRecordMember: 258,
        JSDocTypeReference: 259,
        JSDocOptionalType: 260,
        JSDocFunctionType: 261,
        JSDocVariadicType: 262,
        JSDocConstructorType: 263,
        JSDocThisType: 264,
        JSDocComment: 265,
        JSDocTag: 266,
        JSDocParameterTag: 267,
        JSDocReturnTag: 268,
        JSDocTypeTag: 269,
        JSDocTemplateTag: 270,
        SyntaxList: 271,
        Count: 272,
        FirstAssignment: 56,
        LastAssignment: 68,
        FirstReservedWord: 70,
        LastReservedWord: 105,
        FirstKeyword: 70,
        LastKeyword: 134,
        FirstFutureReservedWord: 106,
        LastFutureReservedWord: 114,
        FirstTypeNode: 151,
        LastTypeNode: 160,
        FirstPunctuation: 15,
        LastPunctuation: 68,
        FirstToken: 0,
        LastToken: 134,
        FirstTriviaToken: 2,
        LastTriviaToken: 7,
        FirstLiteralToken: 8,
        LastLiteralToken: 11,
        FirstTemplateToken: 11,
        LastTemplateToken: 14,
        FirstBinaryOperator: 25,
        LastBinaryOperator: 68,
        FirstNode: 135,
    };
    var SyntaxKind_1_8 = {
        Unknown: 0,
        EndOfFileToken: 1,
        SingleLineCommentTrivia: 2,
        MultiLineCommentTrivia: 3,
        NewLineTrivia: 4,
        WhitespaceTrivia: 5,
        ShebangTrivia: 6,
        ConflictMarkerTrivia: 7,
        NumericLiteral: 8,
        StringLiteral: 9,
        RegularExpressionLiteral: 10,
        NoSubstitutionTemplateLiteral: 11,
        TemplateHead: 12,
        TemplateMiddle: 13,
        TemplateTail: 14,
        OpenBraceToken: 15,
        CloseBraceToken: 16,
        OpenParenToken: 17,
        CloseParenToken: 18,
        OpenBracketToken: 19,
        CloseBracketToken: 20,
        DotToken: 21,
        DotDotDotToken: 22,
        SemicolonToken: 23,
        CommaToken: 24,
        LessThanToken: 25,
        LessThanSlashToken: 26,
        GreaterThanToken: 27,
        LessThanEqualsToken: 28,
        GreaterThanEqualsToken: 29,
        EqualsEqualsToken: 30,
        ExclamationEqualsToken: 31,
        EqualsEqualsEqualsToken: 32,
        ExclamationEqualsEqualsToken: 33,
        EqualsGreaterThanToken: 34,
        PlusToken: 35,
        MinusToken: 36,
        AsteriskToken: 37,
        AsteriskAsteriskToken: 38,
        SlashToken: 39,
        PercentToken: 40,
        PlusPlusToken: 41,
        MinusMinusToken: 42,
        LessThanLessThanToken: 43,
        GreaterThanGreaterThanToken: 44,
        GreaterThanGreaterThanGreaterThanToken: 45,
        AmpersandToken: 46,
        BarToken: 47,
        CaretToken: 48,
        ExclamationToken: 49,
        TildeToken: 50,
        AmpersandAmpersandToken: 51,
        BarBarToken: 52,
        QuestionToken: 53,
        ColonToken: 54,
        AtToken: 55,
        EqualsToken: 56,
        PlusEqualsToken: 57,
        MinusEqualsToken: 58,
        AsteriskEqualsToken: 59,
        AsteriskAsteriskEqualsToken: 60,
        SlashEqualsToken: 61,
        PercentEqualsToken: 62,
        LessThanLessThanEqualsToken: 63,
        GreaterThanGreaterThanEqualsToken: 64,
        GreaterThanGreaterThanGreaterThanEqualsToken: 65,
        AmpersandEqualsToken: 66,
        BarEqualsToken: 67,
        CaretEqualsToken: 68,
        Identifier: 69,
        BreakKeyword: 70,
        CaseKeyword: 71,
        CatchKeyword: 72,
        ClassKeyword: 73,
        ConstKeyword: 74,
        ContinueKeyword: 75,
        DebuggerKeyword: 76,
        DefaultKeyword: 77,
        DeleteKeyword: 78,
        DoKeyword: 79,
        ElseKeyword: 80,
        EnumKeyword: 81,
        ExportKeyword: 82,
        ExtendsKeyword: 83,
        FalseKeyword: 84,
        FinallyKeyword: 85,
        ForKeyword: 86,
        FunctionKeyword: 87,
        IfKeyword: 88,
        ImportKeyword: 89,
        InKeyword: 90,
        InstanceOfKeyword: 91,
        NewKeyword: 92,
        NullKeyword: 93,
        ReturnKeyword: 94,
        SuperKeyword: 95,
        SwitchKeyword: 96,
        ThisKeyword: 97,
        ThrowKeyword: 98,
        TrueKeyword: 99,
        TryKeyword: 100,
        TypeOfKeyword: 101,
        VarKeyword: 102,
        VoidKeyword: 103,
        WhileKeyword: 104,
        WithKeyword: 105,
        ImplementsKeyword: 106,
        InterfaceKeyword: 107,
        LetKeyword: 108,
        PackageKeyword: 109,
        PrivateKeyword: 110,
        ProtectedKeyword: 111,
        PublicKeyword: 112,
        StaticKeyword: 113,
        YieldKeyword: 114,
        AbstractKeyword: 115,
        AsKeyword: 116,
        AnyKeyword: 117,
        AsyncKeyword: 118,
        AwaitKeyword: 119,
        BooleanKeyword: 120,
        ConstructorKeyword: 121,
        DeclareKeyword: 122,
        GetKeyword: 123,
        IsKeyword: 124,
        ModuleKeyword: 125,
        NamespaceKeyword: 126,
        RequireKeyword: 127,
        NumberKeyword: 128,
        SetKeyword: 129,
        StringKeyword: 130,
        SymbolKeyword: 131,
        TypeKeyword: 132,
        FromKeyword: 133,
        GlobalKeyword: 134,
        OfKeyword: 135,
        QualifiedName: 136,
        ComputedPropertyName: 137,
        TypeParameter: 138,
        Parameter: 139,
        Decorator: 140,
        PropertySignature: 141,
        PropertyDeclaration: 142,
        MethodSignature: 143,
        MethodDeclaration: 144,
        Constructor: 145,
        GetAccessor: 146,
        SetAccessor: 147,
        CallSignature: 148,
        ConstructSignature: 149,
        IndexSignature: 150,
        TypePredicate: 151,
        TypeReference: 152,
        FunctionType: 153,
        ConstructorType: 154,
        TypeQuery: 155,
        TypeLiteral: 156,
        ArrayType: 157,
        TupleType: 158,
        UnionType: 159,
        IntersectionType: 160,
        ParenthesizedType: 161,
        ThisType: 162,
        StringLiteralType: 163,
        ObjectBindingPattern: 164,
        ArrayBindingPattern: 165,
        BindingElement: 166,
        ArrayLiteralExpression: 167,
        ObjectLiteralExpression: 168,
        PropertyAccessExpression: 169,
        ElementAccessExpression: 170,
        CallExpression: 171,
        NewExpression: 172,
        TaggedTemplateExpression: 173,
        TypeAssertionExpression: 174,
        ParenthesizedExpression: 175,
        FunctionExpression: 176,
        ArrowFunction: 177,
        DeleteExpression: 178,
        TypeOfExpression: 179,
        VoidExpression: 180,
        AwaitExpression: 181,
        PrefixUnaryExpression: 182,
        PostfixUnaryExpression: 183,
        BinaryExpression: 184,
        ConditionalExpression: 185,
        TemplateExpression: 186,
        YieldExpression: 187,
        SpreadElementExpression: 188,
        ClassExpression: 189,
        OmittedExpression: 190,
        ExpressionWithTypeArguments: 191,
        AsExpression: 192,
        TemplateSpan: 193,
        SemicolonClassElement: 194,
        Block: 195,
        VariableStatement: 196,
        EmptyStatement: 197,
        ExpressionStatement: 198,
        IfStatement: 199,
        DoStatement: 200,
        WhileStatement: 201,
        ForStatement: 202,
        ForInStatement: 203,
        ForOfStatement: 204,
        ContinueStatement: 205,
        BreakStatement: 206,
        ReturnStatement: 207,
        WithStatement: 208,
        SwitchStatement: 209,
        LabeledStatement: 210,
        ThrowStatement: 211,
        TryStatement: 212,
        DebuggerStatement: 213,
        VariableDeclaration: 214,
        VariableDeclarationList: 215,
        FunctionDeclaration: 216,
        ClassDeclaration: 217,
        InterfaceDeclaration: 218,
        TypeAliasDeclaration: 219,
        EnumDeclaration: 220,
        ModuleDeclaration: 221,
        ModuleBlock: 222,
        CaseBlock: 223,
        ImportEqualsDeclaration: 224,
        ImportDeclaration: 225,
        ImportClause: 226,
        NamespaceImport: 227,
        NamedImports: 228,
        ImportSpecifier: 229,
        ExportAssignment: 230,
        ExportDeclaration: 231,
        NamedExports: 232,
        ExportSpecifier: 233,
        MissingDeclaration: 234,
        ExternalModuleReference: 235,
        JsxElement: 236,
        JsxSelfClosingElement: 237,
        JsxOpeningElement: 238,
        JsxText: 239,
        JsxClosingElement: 240,
        JsxAttribute: 241,
        JsxSpreadAttribute: 242,
        JsxExpression: 243,
        CaseClause: 244,
        DefaultClause: 245,
        HeritageClause: 246,
        CatchClause: 247,
        PropertyAssignment: 248,
        ShorthandPropertyAssignment: 249,
        EnumMember: 250,
        SourceFile: 251,
        JSDocTypeExpression: 252,
        JSDocAllType: 253,
        JSDocUnknownType: 254,
        JSDocArrayType: 255,
        JSDocUnionType: 256,
        JSDocTupleType: 257,
        JSDocNullableType: 258,
        JSDocNonNullableType: 259,
        JSDocRecordType: 260,
        JSDocRecordMember: 261,
        JSDocTypeReference: 262,
        JSDocOptionalType: 263,
        JSDocFunctionType: 264,
        JSDocVariadicType: 265,
        JSDocConstructorType: 266,
        JSDocThisType: 267,
        JSDocComment: 268,
        JSDocTag: 269,
        JSDocParameterTag: 270,
        JSDocReturnTag: 271,
        JSDocTypeTag: 272,
        JSDocTemplateTag: 273,
        SyntaxList: 274,
        Count: 275,
        FirstAssignment: 56,
        LastAssignment: 68,
        FirstReservedWord: 70,
        LastReservedWord: 105,
        FirstKeyword: 70,
        LastKeyword: 135,
        FirstFutureReservedWord: 106,
        LastFutureReservedWord: 114,
        FirstTypeNode: 151,
        LastTypeNode: 163,
        FirstPunctuation: 15,
        LastPunctuation: 68,
        FirstToken: 0,
        LastToken: 135,
        FirstTriviaToken: 2,
        LastTriviaToken: 7,
        FirstLiteralToken: 8,
        LastLiteralToken: 11,
        FirstTemplateToken: 11,
        LastTemplateToken: 14,
        FirstBinaryOperator: 25,
        LastBinaryOperator: 68,
        FirstNode: 136
    };
    var SyntaxKind_1_9 = {
        Unknown: 0,
        EndOfFileToken: 1,
        SingleLineCommentTrivia: 2,
        MultiLineCommentTrivia: 3,
        NewLineTrivia: 4,
        WhitespaceTrivia: 5,
        ShebangTrivia: 6,
        ConflictMarkerTrivia: 7,
        NumericLiteral: 8,
        StringLiteral: 9,
        RegularExpressionLiteral: 10,
        NoSubstitutionTemplateLiteral: 11,
        TemplateHead: 12,
        TemplateMiddle: 13,
        TemplateTail: 14,
        OpenBraceToken: 15,
        CloseBraceToken: 16,
        OpenParenToken: 17,
        CloseParenToken: 18,
        OpenBracketToken: 19,
        CloseBracketToken: 20,
        DotToken: 21,
        DotDotDotToken: 22,
        SemicolonToken: 23,
        CommaToken: 24,
        LessThanToken: 25,
        LessThanSlashToken: 26,
        GreaterThanToken: 27,
        LessThanEqualsToken: 28,
        GreaterThanEqualsToken: 29,
        EqualsEqualsToken: 30,
        ExclamationEqualsToken: 31,
        EqualsEqualsEqualsToken: 32,
        ExclamationEqualsEqualsToken: 33,
        EqualsGreaterThanToken: 34,
        PlusToken: 35,
        MinusToken: 36,
        AsteriskToken: 37,
        AsteriskAsteriskToken: 38,
        SlashToken: 39,
        PercentToken: 40,
        PlusPlusToken: 41,
        MinusMinusToken: 42,
        LessThanLessThanToken: 43,
        GreaterThanGreaterThanToken: 44,
        GreaterThanGreaterThanGreaterThanToken: 45,
        AmpersandToken: 46,
        BarToken: 47,
        CaretToken: 48,
        ExclamationToken: 49,
        TildeToken: 50,
        AmpersandAmpersandToken: 51,
        BarBarToken: 52,
        QuestionToken: 53,
        ColonToken: 54,
        AtToken: 55,
        EqualsToken: 56,
        PlusEqualsToken: 57,
        MinusEqualsToken: 58,
        AsteriskEqualsToken: 59,
        AsteriskAsteriskEqualsToken: 60,
        SlashEqualsToken: 61,
        PercentEqualsToken: 62,
        LessThanLessThanEqualsToken: 63,
        GreaterThanGreaterThanEqualsToken: 64,
        GreaterThanGreaterThanGreaterThanEqualsToken: 65,
        AmpersandEqualsToken: 66,
        BarEqualsToken: 67,
        CaretEqualsToken: 68,
        Identifier: 69,
        BreakKeyword: 70,
        CaseKeyword: 71,
        CatchKeyword: 72,
        ClassKeyword: 73,
        ConstKeyword: 74,
        ContinueKeyword: 75,
        DebuggerKeyword: 76,
        DefaultKeyword: 77,
        DeleteKeyword: 78,
        DoKeyword: 79,
        ElseKeyword: 80,
        EnumKeyword: 81,
        ExportKeyword: 82,
        ExtendsKeyword: 83,
        FalseKeyword: 84,
        FinallyKeyword: 85,
        ForKeyword: 86,
        FunctionKeyword: 87,
        IfKeyword: 88,
        ImportKeyword: 89,
        InKeyword: 90,
        InstanceOfKeyword: 91,
        NewKeyword: 92,
        NullKeyword: 93,
        ReturnKeyword: 94,
        SuperKeyword: 95,
        SwitchKeyword: 96,
        ThisKeyword: 97,
        ThrowKeyword: 98,
        TrueKeyword: 99,
        TryKeyword: 100,
        TypeOfKeyword: 101,
        VarKeyword: 102,
        VoidKeyword: 103,
        WhileKeyword: 104,
        WithKeyword: 105,
        ImplementsKeyword: 106,
        InterfaceKeyword: 107,
        LetKeyword: 108,
        PackageKeyword: 109,
        PrivateKeyword: 110,
        ProtectedKeyword: 111,
        PublicKeyword: 112,
        StaticKeyword: 113,
        YieldKeyword: 114,
        AbstractKeyword: 115,
        AsKeyword: 116,
        AnyKeyword: 117,
        AsyncKeyword: 118,
        AwaitKeyword: 119,
        BooleanKeyword: 120,
        ConstructorKeyword: 121,
        DeclareKeyword: 122,
        GetKeyword: 123,
        IsKeyword: 124,
        ModuleKeyword: 125,
        NamespaceKeyword: 126,
        RequireKeyword: 128,
        NumberKeyword: 129,
        SetKeyword: 130,
        StringKeyword: 131,
        SymbolKeyword: 132,
        TypeKeyword: 133,
        FromKeyword: 135,
        GlobalKeyword: 136,
        OfKeyword: 137,
        QualifiedName: 138,
        ComputedPropertyName: 139,
        TypeParameter: 140,
        Parameter: 141,
        Decorator: 142,
        PropertySignature: 143,
        PropertyDeclaration: 144,
        MethodSignature: 145,
        MethodDeclaration: 146,
        Constructor: 147,
        GetAccessor: 148,
        SetAccessor: 149,
        CallSignature: 150,
        ConstructSignature: 151,
        IndexSignature: 152,
        TypePredicate: 153,
        TypeReference: 154,
        FunctionType: 155,
        ConstructorType: 156,
        TypeQuery: 157,
        TypeLiteral: 158,
        ArrayType: 159,
        TupleType: 160,
        UnionType: 161,
        IntersectionType: 162,
        ParenthesizedType: 163,
        ThisType: 164,
        StringLiteralType: 165,
        ObjectBindingPattern: 166,
        ArrayBindingPattern: 167,
        BindingElement: 168,
        ArrayLiteralExpression: 169,
        ObjectLiteralExpression: 170,
        PropertyAccessExpression: 171,
        ElementAccessExpression: 172,
        CallExpression: 173,
        NewExpression: 174,
        TaggedTemplateExpression: 175,
        TypeAssertionExpression: 176,
        ParenthesizedExpression: 177,
        FunctionExpression: 178,
        ArrowFunction: 179,
        DeleteExpression: 180,
        TypeOfExpression: 181,
        VoidExpression: 182,
        AwaitExpression: 183,
        PrefixUnaryExpression: 184,
        PostfixUnaryExpression: 185,
        BinaryExpression: 186,
        ConditionalExpression: 187,
        TemplateExpression: 188,
        YieldExpression: 189,
        SpreadElementExpression: 190,
        ClassExpression: 191,
        OmittedExpression: 192,
        ExpressionWithTypeArguments: 193,
        AsExpression: 194,
        TemplateSpan: 196,
        SemicolonClassElement: 197,
        Block: 198,
        VariableStatement: 199,
        EmptyStatement: 200,
        ExpressionStatement: 201,
        IfStatement: 202,
        DoStatement: 203,
        WhileStatement: 204,
        ForStatement: 205,
        ForInStatement: 206,
        ForOfStatement: 207,
        ContinueStatement: 208,
        BreakStatement: 209,
        ReturnStatement: 210,
        WithStatement: 211,
        SwitchStatement: 212,
        LabeledStatement: 213,
        ThrowStatement: 214,
        TryStatement: 215,
        DebuggerStatement: 216,
        VariableDeclaration: 217,
        VariableDeclarationList: 218,
        FunctionDeclaration: 219,
        ClassDeclaration: 220,
        InterfaceDeclaration: 221,
        TypeAliasDeclaration: 222,
        EnumDeclaration: 223,
        ModuleDeclaration: 224,
        ModuleBlock: 225,
        CaseBlock: 226,
        ImportEqualsDeclaration: 228,
        ImportDeclaration: 229,
        ImportClause: 230,
        NamespaceImport: 231,
        NamedImports: 232,
        ImportSpecifier: 233,
        ExportAssignment: 234,
        ExportDeclaration: 235,
        NamedExports: 236,
        ExportSpecifier: 237,
        MissingDeclaration: 238,
        ExternalModuleReference: 239,
        JsxElement: 240,
        JsxSelfClosingElement: 241,
        JsxOpeningElement: 242,
        JsxText: 243,
        JsxClosingElement: 244,
        JsxAttribute: 245,
        JsxSpreadAttribute: 246,
        JsxExpression: 247,
        CaseClause: 248,
        DefaultClause: 249,
        HeritageClause: 250,
        CatchClause: 251,
        PropertyAssignment: 252,
        ShorthandPropertyAssignment: 253,
        EnumMember: 254,
        SourceFile: 255,
        JSDocTypeExpression: 256,
        JSDocAllType: 257,
        JSDocUnknownType: 258,
        JSDocArrayType: 259,
        JSDocUnionType: 260,
        JSDocTupleType: 261,
        JSDocNullableType: 262,
        JSDocNonNullableType: 263,
        JSDocRecordType: 264,
        JSDocRecordMember: 265,
        JSDocTypeReference: 266,
        JSDocOptionalType: 267,
        JSDocFunctionType: 268,
        JSDocVariadicType: 269,
        JSDocConstructorType: 270,
        JSDocThisType: 271,
        JSDocComment: 272,
        JSDocTag: 273,
        JSDocParameterTag: 274,
        JSDocReturnTag: 275,
        JSDocTypeTag: 276,
        JSDocTemplateTag: 277,
        SyntaxList: 278,
        Count: 279,
        FirstAssignment: 56,
        LastAssignment: 68,
        FirstReservedWord: 70,
        LastReservedWord: 105,
        FirstKeyword: 70,
        LastKeyword: 137,
        FirstFutureReservedWord: 106,
        LastFutureReservedWord: 114,
        FirstTypeNode: 153,
        LastTypeNode: 165,
        FirstPunctuation: 15,
        LastPunctuation: 68,
        FirstToken: 0,
        LastToken: 137,
        FirstTriviaToken: 2,
        LastTriviaToken: 7,
        FirstLiteralToken: 8,
        LastLiteralToken: 11,
        FirstTemplateToken: 11,
        LastTemplateToken: 14,
        FirstBinaryOperator: 25,
        LastBinaryOperator: 68,
        FirstNode: 138,
        GlobalModuleExportDeclaration: 227,
        NonNullExpression: 195,
        UndefinedKeyword: 134,
        ReadonlyKeyword: 127
    };
    var SyntaxKind_2_0 = {
        Unknown: 0,
        EndOfFileToken: 1,
        SingleLineCommentTrivia: 2,
        MultiLineCommentTrivia: 3,
        NewLineTrivia: 4,
        WhitespaceTrivia: 5,
        ShebangTrivia: 6,
        ConflictMarkerTrivia: 7,
        NumericLiteral: 8,
        StringLiteral: 9,
        RegularExpressionLiteral: 10,
        NoSubstitutionTemplateLiteral: 11,
        TemplateHead: 12,
        TemplateMiddle: 13,
        TemplateTail: 14,
        OpenBraceToken: 15,
        CloseBraceToken: 16,
        OpenParenToken: 17,
        CloseParenToken: 18,
        OpenBracketToken: 19,
        CloseBracketToken: 20,
        DotToken: 21,
        DotDotDotToken: 22,
        SemicolonToken: 23,
        CommaToken: 24,
        LessThanToken: 25,
        LessThanSlashToken: 26,
        GreaterThanToken: 27,
        LessThanEqualsToken: 28,
        GreaterThanEqualsToken: 29,
        EqualsEqualsToken: 30,
        ExclamationEqualsToken: 31,
        EqualsEqualsEqualsToken: 32,
        ExclamationEqualsEqualsToken: 33,
        EqualsGreaterThanToken: 34,
        PlusToken: 35,
        MinusToken: 36,
        AsteriskToken: 37,
        AsteriskAsteriskToken: 38,
        SlashToken: 39,
        PercentToken: 40,
        PlusPlusToken: 41,
        MinusMinusToken: 42,
        LessThanLessThanToken: 43,
        GreaterThanGreaterThanToken: 44,
        GreaterThanGreaterThanGreaterThanToken: 45,
        AmpersandToken: 46,
        BarToken: 47,
        CaretToken: 48,
        ExclamationToken: 49,
        TildeToken: 50,
        AmpersandAmpersandToken: 51,
        BarBarToken: 52,
        QuestionToken: 53,
        ColonToken: 54,
        AtToken: 55,
        EqualsToken: 56,
        PlusEqualsToken: 57,
        MinusEqualsToken: 58,
        AsteriskEqualsToken: 59,
        AsteriskAsteriskEqualsToken: 60,
        SlashEqualsToken: 61,
        PercentEqualsToken: 62,
        LessThanLessThanEqualsToken: 63,
        GreaterThanGreaterThanEqualsToken: 64,
        GreaterThanGreaterThanGreaterThanEqualsToken: 65,
        AmpersandEqualsToken: 66,
        BarEqualsToken: 67,
        CaretEqualsToken: 68,
        Identifier: 69,
        BreakKeyword: 70,
        CaseKeyword: 71,
        CatchKeyword: 72,
        ClassKeyword: 73,
        ConstKeyword: 74,
        ContinueKeyword: 75,
        DebuggerKeyword: 76,
        DefaultKeyword: 77,
        DeleteKeyword: 78,
        DoKeyword: 79,
        ElseKeyword: 80,
        EnumKeyword: 81,
        ExportKeyword: 82,
        ExtendsKeyword: 83,
        FalseKeyword: 84,
        FinallyKeyword: 85,
        ForKeyword: 86,
        FunctionKeyword: 87,
        IfKeyword: 88,
        ImportKeyword: 89,
        InKeyword: 90,
        InstanceOfKeyword: 91,
        NewKeyword: 92,
        NullKeyword: 93,
        ReturnKeyword: 94,
        SuperKeyword: 95,
        SwitchKeyword: 96,
        ThisKeyword: 97,
        ThrowKeyword: 98,
        TrueKeyword: 99,
        TryKeyword: 100,
        TypeOfKeyword: 101,
        VarKeyword: 102,
        VoidKeyword: 103,
        WhileKeyword: 104,
        WithKeyword: 105,
        ImplementsKeyword: 106,
        InterfaceKeyword: 107,
        LetKeyword: 108,
        PackageKeyword: 109,
        PrivateKeyword: 110,
        ProtectedKeyword: 111,
        PublicKeyword: 112,
        StaticKeyword: 113,
        YieldKeyword: 114,
        AbstractKeyword: 115,
        AsKeyword: 116,
        AnyKeyword: 117,
        AsyncKeyword: 118,
        AwaitKeyword: 119,
        BooleanKeyword: 120,
        ConstructorKeyword: 121,
        DeclareKeyword: 122,
        GetKeyword: 123,
        IsKeyword: 124,
        ModuleKeyword: 125,
        NamespaceKeyword: 126,
        NeverKeyword: 127,
        ReadonlyKeyword: 128,
        RequireKeyword: 129,
        NumberKeyword: 130,
        SetKeyword: 131,
        StringKeyword: 132,
        SymbolKeyword: 133,
        TypeKeyword: 134,
        UndefinedKeyword: 135,
        FromKeyword: 136,
        GlobalKeyword: 137,
        OfKeyword: 138,
        QualifiedName: 139,
        ComputedPropertyName: 140,
        TypeParameter: 141,
        Parameter: 142,
        Decorator: 143,
        PropertySignature: 144,
        PropertyDeclaration: 145,
        MethodSignature: 146,
        MethodDeclaration: 147,
        Constructor: 148,
        GetAccessor: 149,
        SetAccessor: 150,
        CallSignature: 151,
        ConstructSignature: 152,
        IndexSignature: 153,
        TypePredicate: 154,
        TypeReference: 155,
        FunctionType: 156,
        ConstructorType: 157,
        TypeQuery: 158,
        TypeLiteral: 159,
        ArrayType: 160,
        TupleType: 161,
        UnionType: 162,
        IntersectionType: 163,
        ParenthesizedType: 164,
        ThisType: 165,
        StringLiteralType: 166,
        ObjectBindingPattern: 167,
        ArrayBindingPattern: 168,
        BindingElement: 169,
        ArrayLiteralExpression: 170,
        ObjectLiteralExpression: 171,
        PropertyAccessExpression: 172,
        ElementAccessExpression: 173,
        CallExpression: 174,
        NewExpression: 175,
        TaggedTemplateExpression: 176,
        TypeAssertionExpression: 177,
        ParenthesizedExpression: 178,
        FunctionExpression: 179,
        ArrowFunction: 180,
        DeleteExpression: 181,
        TypeOfExpression: 182,
        VoidExpression: 183,
        AwaitExpression: 184,
        PrefixUnaryExpression: 185,
        PostfixUnaryExpression: 186,
        BinaryExpression: 187,
        ConditionalExpression: 188,
        TemplateExpression: 189,
        YieldExpression: 190,
        SpreadElementExpression: 191,
        ClassExpression: 192,
        OmittedExpression: 193,
        ExpressionWithTypeArguments: 194,
        AsExpression: 195,
        NonNullExpression: 196,
        TemplateSpan: 197,
        SemicolonClassElement: 198,
        Block: 199,
        VariableStatement: 200,
        EmptyStatement: 201,
        ExpressionStatement: 202,
        IfStatement: 203,
        DoStatement: 204,
        WhileStatement: 205,
        ForStatement: 206,
        ForInStatement: 207,
        ForOfStatement: 208,
        ContinueStatement: 209,
        BreakStatement: 210,
        ReturnStatement: 211,
        WithStatement: 212,
        SwitchStatement: 213,
        LabeledStatement: 214,
        ThrowStatement: 215,
        TryStatement: 216,
        DebuggerStatement: 217,
        VariableDeclaration: 218,
        VariableDeclarationList: 219,
        FunctionDeclaration: 220,
        ClassDeclaration: 221,
        InterfaceDeclaration: 222,
        TypeAliasDeclaration: 223,
        EnumDeclaration: 224,
        ModuleDeclaration: 225,
        ModuleBlock: 226,
        CaseBlock: 227,
        NamespaceExportDeclaration: 228,
        ImportEqualsDeclaration: 229,
        ImportDeclaration: 230,
        ImportClause: 231,
        NamespaceImport: 232,
        NamedImports: 233,
        ImportSpecifier: 234,
        ExportAssignment: 235,
        ExportDeclaration: 236,
        NamedExports: 237,
        ExportSpecifier: 238,
        MissingDeclaration: 239,
        ExternalModuleReference: 240,
        JsxElement: 241,
        JsxSelfClosingElement: 242,
        JsxOpeningElement: 243,
        JsxText: 244,
        JsxClosingElement: 245,
        JsxAttribute: 246,
        JsxSpreadAttribute: 247,
        JsxExpression: 248,
        CaseClause: 249,
        DefaultClause: 250,
        HeritageClause: 251,
        CatchClause: 252,
        PropertyAssignment: 253,
        ShorthandPropertyAssignment: 254,
        EnumMember: 255,
        SourceFile: 256,
        JSDocTypeExpression: 257,
        JSDocAllType: 258,
        JSDocUnknownType: 259,
        JSDocArrayType: 260,
        JSDocUnionType: 261,
        JSDocTupleType: 262,
        JSDocNullableType: 263,
        JSDocNonNullableType: 264,
        JSDocRecordType: 265,
        JSDocRecordMember: 266,
        JSDocTypeReference: 267,
        JSDocOptionalType: 268,
        JSDocFunctionType: 269,
        JSDocVariadicType: 270,
        JSDocConstructorType: 271,
        JSDocThisType: 272,
        JSDocComment: 273,
        JSDocTag: 274,
        JSDocParameterTag: 275,
        JSDocReturnTag: 276,
        JSDocTypeTag: 277,
        JSDocTemplateTag: 278,
        JSDocTypedefTag: 279,
        JSDocPropertyTag: 280,
        JSDocTypeLiteral: 281,
        SyntaxList: 282,
        Count: 283,
        FirstAssignment: 56,
        LastAssignment: 68,
        FirstReservedWord: 70,
        LastReservedWord: 105,
        FirstKeyword: 70,
        LastKeyword: 138,
        FirstFutureReservedWord: 106,
        LastFutureReservedWord: 114,
        FirstTypeNode: 154,
        LastTypeNode: 166,
        FirstPunctuation: 15,
        LastPunctuation: 68,
        FirstToken: 0,
        LastToken: 138,
        FirstTriviaToken: 2,
        LastTriviaToken: 7,
        FirstLiteralToken: 8,
        LastLiteralToken: 11,
        FirstTemplateToken: 11,
        LastTemplateToken: 14,
        FirstBinaryOperator: 25,
        LastBinaryOperator: 68,
        FirstNode: 139,
        FirstJSDocNode: 257,
        LastJSDocNode: 281,
        FirstJSDocTagNode: 273,
        LastJSDocTagNode: 281,
    };
    var SyntaxKind_2_1 = {
        Unknown: 0,
        EndOfFileToken: 1,
        SingleLineCommentTrivia: 2,
        MultiLineCommentTrivia: 3,
        NewLineTrivia: 4,
        WhitespaceTrivia: 5,
        ShebangTrivia: 6,
        ConflictMarkerTrivia: 7,
        NumericLiteral: 8,
        StringLiteral: 9,
        JsxText: 10,
        RegularExpressionLiteral: 11,
        NoSubstitutionTemplateLiteral: 12,
        TemplateHead: 13,
        TemplateMiddle: 14,
        TemplateTail: 15,
        OpenBraceToken: 16,
        CloseBraceToken: 17,
        OpenParenToken: 18,
        CloseParenToken: 19,
        OpenBracketToken: 20,
        CloseBracketToken: 21,
        DotToken: 22,
        DotDotDotToken: 23,
        SemicolonToken: 24,
        CommaToken: 25,
        LessThanToken: 26,
        LessThanSlashToken: 27,
        GreaterThanToken: 28,
        LessThanEqualsToken: 29,
        GreaterThanEqualsToken: 30,
        EqualsEqualsToken: 31,
        ExclamationEqualsToken: 32,
        EqualsEqualsEqualsToken: 33,
        ExclamationEqualsEqualsToken: 34,
        EqualsGreaterThanToken: 35,
        PlusToken: 36,
        MinusToken: 37,
        AsteriskToken: 38,
        AsteriskAsteriskToken: 39,
        SlashToken: 40,
        PercentToken: 41,
        PlusPlusToken: 42,
        MinusMinusToken: 43,
        LessThanLessThanToken: 44,
        GreaterThanGreaterThanToken: 45,
        GreaterThanGreaterThanGreaterThanToken: 46,
        AmpersandToken: 47,
        BarToken: 48,
        CaretToken: 49,
        ExclamationToken: 50,
        TildeToken: 51,
        AmpersandAmpersandToken: 52,
        BarBarToken: 53,
        QuestionToken: 54,
        ColonToken: 55,
        AtToken: 56,
        EqualsToken: 57,
        PlusEqualsToken: 58,
        MinusEqualsToken: 59,
        AsteriskEqualsToken: 60,
        AsteriskAsteriskEqualsToken: 61,
        SlashEqualsToken: 62,
        PercentEqualsToken: 63,
        LessThanLessThanEqualsToken: 64,
        GreaterThanGreaterThanEqualsToken: 65,
        GreaterThanGreaterThanGreaterThanEqualsToken: 66,
        AmpersandEqualsToken: 67,
        BarEqualsToken: 68,
        CaretEqualsToken: 69,
        Identifier: 70,
        BreakKeyword: 71,
        CaseKeyword: 72,
        CatchKeyword: 73,
        ClassKeyword: 74,
        ConstKeyword: 75,
        ContinueKeyword: 76,
        DebuggerKeyword: 77,
        DefaultKeyword: 78,
        DeleteKeyword: 79,
        DoKeyword: 80,
        ElseKeyword: 81,
        EnumKeyword: 82,
        ExportKeyword: 83,
        ExtendsKeyword: 84,
        FalseKeyword: 85,
        FinallyKeyword: 86,
        ForKeyword: 87,
        FunctionKeyword: 88,
        IfKeyword: 89,
        ImportKeyword: 90,
        InKeyword: 91,
        InstanceOfKeyword: 92,
        NewKeyword: 93,
        NullKeyword: 94,
        ReturnKeyword: 95,
        SuperKeyword: 96,
        SwitchKeyword: 97,
        ThisKeyword: 98,
        ThrowKeyword: 99,
        TrueKeyword: 100,
        TryKeyword: 101,
        TypeOfKeyword: 102,
        VarKeyword: 103,
        VoidKeyword: 104,
        WhileKeyword: 105,
        WithKeyword: 106,
        ImplementsKeyword: 107,
        InterfaceKeyword: 108,
        LetKeyword: 109,
        PackageKeyword: 110,
        PrivateKeyword: 111,
        ProtectedKeyword: 112,
        PublicKeyword: 113,
        StaticKeyword: 114,
        YieldKeyword: 115,
        AbstractKeyword: 116,
        AsKeyword: 117,
        AnyKeyword: 118,
        AsyncKeyword: 119,
        AwaitKeyword: 120,
        BooleanKeyword: 121,
        ConstructorKeyword: 122,
        DeclareKeyword: 123,
        GetKeyword: 124,
        IsKeyword: 125,
        ModuleKeyword: 126,
        NamespaceKeyword: 127,
        NeverKeyword: 128,
        ReadonlyKeyword: 129,
        RequireKeyword: 130,
        NumberKeyword: 131,
        SetKeyword: 132,
        StringKeyword: 133,
        SymbolKeyword: 134,
        TypeKeyword: 135,
        UndefinedKeyword: 136,
        FromKeyword: 137,
        GlobalKeyword: 138,
        OfKeyword: 139,
        QualifiedName: 140,
        ComputedPropertyName: 141,
        TypeParameter: 142,
        Parameter: 143,
        Decorator: 144,
        PropertySignature: 145,
        PropertyDeclaration: 146,
        MethodSignature: 147,
        MethodDeclaration: 148,
        Constructor: 149,
        GetAccessor: 150,
        SetAccessor: 151,
        CallSignature: 152,
        ConstructSignature: 153,
        IndexSignature: 154,
        TypePredicate: 155,
        TypeReference: 156,
        FunctionType: 157,
        ConstructorType: 158,
        TypeQuery: 159,
        TypeLiteral: 160,
        ArrayType: 161,
        TupleType: 162,
        UnionType: 163,
        IntersectionType: 164,
        ParenthesizedType: 165,
        ThisType: 166,
        LiteralType: 167,
        ObjectBindingPattern: 168,
        ArrayBindingPattern: 169,
        BindingElement: 170,
        ArrayLiteralExpression: 171,
        ObjectLiteralExpression: 172,
        PropertyAccessExpression: 173,
        ElementAccessExpression: 174,
        CallExpression: 175,
        NewExpression: 176,
        TaggedTemplateExpression: 177,
        TypeAssertionExpression: 178,
        ParenthesizedExpression: 179,
        FunctionExpression: 180,
        ArrowFunction: 181,
        DeleteExpression: 182,
        TypeOfExpression: 183,
        VoidExpression: 184,
        AwaitExpression: 185,
        PrefixUnaryExpression: 186,
        PostfixUnaryExpression: 187,
        BinaryExpression: 188,
        ConditionalExpression: 189,
        TemplateExpression: 190,
        YieldExpression: 191,
        SpreadElementExpression: 192,
        ClassExpression: 193,
        OmittedExpression: 194,
        ExpressionWithTypeArguments: 195,
        AsExpression: 196,
        NonNullExpression: 197,
        TemplateSpan: 198,
        SemicolonClassElement: 199,
        Block: 200,
        VariableStatement: 201,
        EmptyStatement: 202,
        ExpressionStatement: 203,
        IfStatement: 204,
        DoStatement: 205,
        WhileStatement: 206,
        ForStatement: 207,
        ForInStatement: 208,
        ForOfStatement: 209,
        ContinueStatement: 210,
        BreakStatement: 211,
        ReturnStatement: 212,
        WithStatement: 213,
        SwitchStatement: 214,
        LabeledStatement: 215,
        ThrowStatement: 216,
        TryStatement: 217,
        DebuggerStatement: 218,
        VariableDeclaration: 219,
        VariableDeclarationList: 220,
        FunctionDeclaration: 221,
        ClassDeclaration: 222,
        InterfaceDeclaration: 223,
        TypeAliasDeclaration: 224,
        EnumDeclaration: 225,
        ModuleDeclaration: 226,
        ModuleBlock: 227,
        CaseBlock: 228,
        NamespaceExportDeclaration: 229,
        ImportEqualsDeclaration: 230,
        ImportDeclaration: 231,
        ImportClause: 232,
        NamespaceImport: 233,
        NamedImports: 234,
        ImportSpecifier: 235,
        ExportAssignment: 236,
        ExportDeclaration: 237,
        NamedExports: 238,
        ExportSpecifier: 239,
        MissingDeclaration: 240,
        ExternalModuleReference: 241,
        JsxElement: 242,
        JsxSelfClosingElement: 243,
        JsxOpeningElement: 244,
        JsxClosingElement: 245,
        JsxAttribute: 246,
        JsxSpreadAttribute: 247,
        JsxExpression: 248,
        CaseClause: 249,
        DefaultClause: 250,
        HeritageClause: 251,
        CatchClause: 252,
        PropertyAssignment: 253,
        ShorthandPropertyAssignment: 254,
        EnumMember: 255,
        SourceFile: 256,
        JSDocTypeExpression: 257,
        JSDocAllType: 258,
        JSDocUnknownType: 259,
        JSDocArrayType: 260,
        JSDocUnionType: 261,
        JSDocTupleType: 262,
        JSDocNullableType: 263,
        JSDocNonNullableType: 264,
        JSDocRecordType: 265,
        JSDocRecordMember: 266,
        JSDocTypeReference: 267,
        JSDocOptionalType: 268,
        JSDocFunctionType: 269,
        JSDocVariadicType: 270,
        JSDocConstructorType: 271,
        JSDocThisType: 272,
        JSDocComment: 273,
        JSDocTag: 274,
        JSDocParameterTag: 275,
        JSDocReturnTag: 276,
        JSDocTypeTag: 277,
        JSDocTemplateTag: 278,
        JSDocTypedefTag: 279,
        JSDocPropertyTag: 280,
        JSDocTypeLiteral: 281,
        JSDocLiteralType: 282,
        JSDocNullKeyword: 283,
        JSDocUndefinedKeyword: 284,
        JSDocNeverKeyword: 285,
        SyntaxList: 286,
        NotEmittedStatement: 287,
        PartiallyEmittedExpression: 288,
        Count: 289,
        FirstAssignment: 57,
        LastAssignment: 69,
        FirstCompoundAssignment: 58,
        LastCompoundAssignment: 69,
        FirstReservedWord: 71,
        LastReservedWord: 106,
        FirstKeyword: 71,
        LastKeyword: 139,
        FirstFutureReservedWord: 107,
        LastFutureReservedWord: 115,
        FirstTypeNode: 155,
        LastTypeNode: 167,
        FirstPunctuation: 16,
        LastPunctuation: 69,
        FirstToken: 0,
        LastToken: 139,
        FirstTriviaToken: 2,
        LastTriviaToken: 7,
        FirstLiteralToken: 8,
        LastLiteralToken: 12,
        FirstTemplateToken: 12,
        LastTemplateToken: 15,
        FirstBinaryOperator: 26,
        LastBinaryOperator: 69,
        FirstNode: 140,
        FirstJSDocNode: 257,
        LastJSDocNode: 282,
        FirstJSDocTagNode: 273,
        LastJSDocTagNode: 285,
    };
    function current() {
        if (cachedKinds === null) {
            if (/1\.5\..*/.test(ts.version)) {
                cachedKinds = SyntaxKind_1_5;
            }
            else if (/1\.6\..*/.test(ts.version)) {
                cachedKinds = SyntaxKind_1_6;
            }
            else if (/1\.7\..*/.test(ts.version)) {
                cachedKinds = SyntaxKind_1_7;
            }
            else if (/1\.8\..*/.test(ts.version)) {
                cachedKinds = SyntaxKind_1_8;
            }
            else if (/1\.9\..*/.test(ts.version)) {
                cachedKinds = SyntaxKind_1_9;
            }
            else if (/2\.0\..*/.test(ts.version)) {
                cachedKinds = SyntaxKind_2_0;
            }
            else if (/2\.1\..*/.test(ts.version)) {
                cachedKinds = SyntaxKind_2_1;
            }
            else {
                throw new Error('Unsupported TypeScript version: ' + ts.version);
            }
        }
        return cachedKinds;
    }
    SyntaxKind.current = current;
})(SyntaxKind || (SyntaxKind = {}));
module.exports = SyntaxKind;
